File: sdk/go/types/node_status.go
// Package types defines data types for the Seocheon SDK.
package types

import "strings"

// NodeStatus represents the status of a registered node.
type NodeStatus string

const (
	NodeStatusUnspecified NodeStatus = "UNSPECIFIED"
	NodeStatusRegistered  NodeStatus = "REGISTERED"
	NodeStatusActive      NodeStatus = "ACTIVE"
	NodeStatusInactive    NodeStatus = "INACTIVE"
	NodeStatusJailed      NodeStatus = "JAILED"
)

// NodeStatusFromInt converts a proto enum integer to NodeStatus string.
func NodeStatusFromInt(status int) NodeStatus {
	switch status {
	case 1:
		return NodeStatusRegistered
	case 2:
		return NodeStatusActive
	case 3:
		return NodeStatusInactive
	case 4:
		return NodeStatusJailed
	default:
		return NodeStatusUnspecified
	}
}

// NodeStatusFromString converts a protobuf JSON status string to NodeStatus.
// Accepts both "NODE_STATUS_REGISTERED" and "REGISTERED" formats.
func NodeStatusFromString(status string) NodeStatus {
	s := strings.TrimPrefix(status, "NODE_STATUS_")
	switch s {
	case "REGISTERED":
		return NodeStatusRegistered
	case "ACTIVE":
		return NodeStatusActive
	case "INACTIVE":
		return NodeStatusInactive
	case "JAILED":
		return NodeStatusJailed
	default:
		return NodeStatusUnspecified
	}
}

File: sdk/csharp/src/Types/NodeStatus.cs
namespace Seocheon.Sdk.Types;

/// <summary>
/// On-chain node status.
/// </summary>
public enum NodeStatus
{
    /// <summary>Registered but not in Active Validator Set.</summary>
    Registered,

    /// <summary>In Active Validator Set.</summary>
    Active,

    /// <summary>Inactive (jailed or unbonded).</summary>
    Inactive
}

/// <summary>
/// Extension methods for NodeStatus.
/// </summary>
public static class NodeStatusExtensions
{
    /// <summary>
    /// Parses a status string from chain response to NodeStatus enum.
    /// </summary>
    public static NodeStatus ParseNodeStatus(string status)
    {
        return status.ToUpperInvariant() switch
        {
            "REGISTERED" => NodeStatus.Registered,
            "ACTIVE" => NodeStatus.Active,
            "INACTIVE" or "JAILED" => NodeStatus.Inactive,
            _ => NodeStatus.Registered,
        };
    }

    /// <summary>
    /// Converts NodeStatus to its on-chain string representation.
    /// </summary>
    public static string ToChainString(this NodeStatus status)
    {
        return status switch
        {
            NodeStatus.Registered => "REGISTERED",
            NodeStatus.Active => "ACTIVE",
            NodeStatus.Inactive => "INACTIVE",
            _ => "REGISTERED",
        };
    }
}
